File: week10/ch.Choe/Week10/Week10/WinnerState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Week10
{
    class WinnerState : State
    {
        GumballMachine m_GumballMachine;

        public WinnerState(GumballMachine machine)
        {
            m_GumballMachine = machine;
        }
        public void Dispense()
        {
            Console.WriteLine("당첨!!");
            m_GumballMachine.ReleaseBall();
            if (m_GumballMachine.Getcount() == 0)
            {
                m_GumballMachine.SetState(GumballMachine.StateType.Soldout);
            }
            else
            {
                m_GumballMachine.ReleaseBall();
                if (m_GumballMachine.Getcount() > 0)
                {
                    m_GumballMachine.SetState(GumballMachine.StateType.NoQuarter);
                }
                else
                {
                    Console.WriteLine("알맹이 없음!!");
                    m_GumballMachine.SetState(GumballMachine.StateType.Soldout);
                }
            }

        }

        public void EjectQuarter()
        {
            Console.WriteLine("이미 뽑았습니다.");
        }

        public void InsertQuarter()
        {
            Console.WriteLine("잠깐 기다려 주세요.");
        }

        public void TurnCrack()
        {
            Console.WriteLine("손잡이는 한번만.");
        }
    }
}

File: week10/minj0i/WinnerState.java
package week10.minj0i;

public class WinnerState implements State {
	GumballMachine2 gumballMachine;
	
	public WinnerState(GumballMachine2 gumballMachine) {
		this.gumballMachine = gumballMachine;
	}
	
	@Override
	public void insertQuarter() {
		System.out.println("오류 메세지");
	}

	@Override
	public void ejectQuarter() {
		System.out.println("오류 메세지");
	}

	@Override
	public void turnCrank() {
		System.out.println("오류 메세지");
	}

	@Override
	public void dispense() {
		System.out.println("축하합니다! 알맹이를 하나 더 받으실 수 있습니다.");
		gumballMachine.releaseBall();
		if (gumballMachine.getCount() == 0) {
			gumballMachine.setState(gumballMachine.getSoldOutState());
		} else {
			gumballMachine.releaseBall();
			if (gumballMachine.getCount() > 0) {
				gumballMachine.setState(gumballMachine.getNoQuarterState());
			} else {
				System.out.println("더 이상 알맹이가 없습니다.");
				gumballMachine.setState(gumballMachine.getSoldOutState());
			}
		}
	}

}
